Code/GraphMol/FileParsers/MultithreadedSDMolSupplier.cpp

```diff
@@ -0,0 +1,262 @@
+//
+//  Copyright (C) 2020 Shrey Aryan
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include "MultithreadedSDMolSupplier.h"
+
+#include "FileParserUtils.h"
+
+namespace RDKit {
+MultithreadedSDMolSupplier::MultithreadedSDMolSupplier(
+    const std::string &fileName, bool sanitize, bool removeHs,
+    bool strictParsing, unsigned int numWriterThreads, size_t sizeInputQueue,
+    size_t sizeOutputQueue) {
+  dp_inStream = openAndCheckStream(fileName);
+  initFromSettings(true, sanitize, removeHs, strictParsing, numWriterThreads,
+                   sizeInputQueue, sizeOutputQueue);
+  POSTCONDITION(dp_inStream, "bad instream");
+  startThreads();
+}
+
+MultithreadedSDMolSupplier::MultithreadedSDMolSupplier(
+    std::istream *inStream, bool takeOwnership, bool sanitize, bool removeHs,
+    bool strictParsing, unsigned int numWriterThreads, size_t sizeInputQueue,
+    size_t sizeOutputQueue) {
+  PRECONDITION(inStream, "bad stream");
+  dp_inStream = inStream;
+  initFromSettings(takeOwnership, sanitize, removeHs, strictParsing,
+                   numWriterThreads, sizeInputQueue, sizeOutputQueue);
+  POSTCONDITION(dp_inStream, "bad instream");
+  startThreads();
+}
+
+MultithreadedSDMolSupplier::MultithreadedSDMolSupplier() {
+  dp_inStream = nullptr;
+  initFromSettings(false, true, true, true, 2, 5, 5);
+  startThreads();
+}
+
+void MultithreadedSDMolSupplier::initFromSettings(bool takeOwnership,
+                                                  bool sanitize, bool removeHs,
+                                                  bool strictParsing,
+                                                  unsigned int numWriterThreads,
+                                                  size_t sizeInputQueue,
+                                                  size_t sizeOutputQueue) {
+  df_owner = takeOwnership;
+  df_sanitize = sanitize;
+  df_removeHs = removeHs;
+  df_strictParsing = strictParsing;
+  d_numWriterThreads = numWriterThreads;
+  d_sizeInputQueue = sizeInputQueue;
+  d_sizeOutputQueue = sizeOutputQueue;
+  d_inputQueue =
+      new ConcurrentQueue<std::tuple<std::string, unsigned int, unsigned int>>(
+          d_sizeInputQueue);
+  d_outputQueue =
+      new ConcurrentQueue<std::tuple<ROMol *, std::string, unsigned int>>(
+          d_sizeOutputQueue);
+
+  df_end = false;
+  d_line = 0;
+  df_processPropertyLists = true;
+}
+
+MultithreadedSDMolSupplier::~MultithreadedSDMolSupplier() {
+  if (df_owner && dp_inStream) {
+    delete dp_inStream;
+    df_owner = false;
+    dp_inStream = nullptr;
+  }
+}
+
+// ensures that there is a line available to be read
+// from the file, implementation identical to the method in
+// in ForwardSDMolSupplier
+void MultithreadedSDMolSupplier::checkForEnd() {
```

Comment:
Instead of copying code in from a method on another class, it would be better to define a helper function and call that from here and `ForwardSDMolSupplier`.

Reply:
The `checkForEnd` here is slightly different from the one implemented in `ForwardSDMolSupplier` since I reset the file pointer after checking for the end of the file (starting from [here](https://github.com/shrey183/rdkit/blob/a9360cf24dabe9712d7b61584fd88497264d5071/Code/GraphMol/FileParsers/MultithreadedSDMolSupplier.cpp#L107)).